File: ui/src/stores/plan.js
import { defineStore } from 'pinia'
import { ref, watchEffect, computed, inject } from 'vue'
import axios from 'axios'
import { deepcopy } from '@/utils/deepcopy'
import { useConfigStore } from '@/stores/config'
import { storeToRefs } from 'pinia'

export const usePlanStore = defineStore('plan', () => {
  const ling_xi = ref(1)
  const exhaust_require = ref([])
  const rest_in_full = ref([])
  const ope_resting_priority = ref([])
  const resting_priority = ref([])
  const workaholic = ref([])
  const refresh_trading = ref([])
  const refresh_drained = ref([])

  const plan = ref({})

  const backup_plans = ref([])

  const operators = ref([])

  const left_side_facility = []

  const facility_operator_limit = {
    central: 5,
    meeting: 2,
    factory: 1,
    contact: 1,
    train: 2
  }
  for (let i = 1; i <= 3; ++i) {
    for (let j = 1; j <= 3; ++j) {
      const facility_name = `room_${i}_${j}`
      const display_name = `B${i}0${j}`
      facility_operator_limit[facility_name] = 3
      left_side_facility.push({ label: display_name, value: facility_name })
    }
  }
  for (let i = 1; i <= 4; ++i) {
    facility_operator_limit[`dormitory_${i}`] = 5
  }
  for (let i = 1; i <= 3; ++i) {
    facility_operator_limit[`gaming_${i}`] = 1
  }

  function list2str(data) {
    return data.join(',')
  }

  function str2list(data) {
    return data && data != '' ? data.split(',') : []
  }

  const backup_conf_convert_list = [
    'exhaust_require',
    'rest_in_full',
    'resting_priority',
    'workaholic',
    'free_blacklist',
    'refresh_trading',
    'refresh_drained',
    'ope_resting_priority'
  ]

  function fill_empty(full_plan) {
    for (const i in facility_operator_limit) {
      let count = 0
      if (!full_plan[i]) {
        count = facility_operator_limit[i]
        full_plan[i] = { name: '', plans: [] }
      } else {
        let limit = facility_operator_limit[i]
        if (full_plan[i].name == '发电站') {
          limit = 1
        } else if (full_plan[i].name == '贸易站') {
          if (!['lmd', 'orundum'].includes(full_plan[i].product)) {
            full_plan[i].product = 'lmd'
          }
        } else if (full_plan[i].name == '制造站') {
          if (!['gold', 'exp3', 'orirock'].includes(full_plan[i].product)) {
            full_plan[i].product = 'gold'
          }
        }
        if (full_plan[i].plans.length < limit) {
          count = limit - full_plan[i].plans.length
        }
      }
      for (let j = 0; j < count; ++j) {
        full_plan[i].plans.push({ agent: '', group: '', replacement: [] })
      }
    }
    return full_plan
  }

  function remove_empty_agent(input) {
    const result = {
      name: input.name,
      plans: []
    }
    if (['贸易站', '制造站'].includes(input.name)) {
      result.product = input.product
    }
    for (const i of input.plans) {
      if (i.agent) {
        result.plans.push(i)
      }
    }
    return result
  }

  function strip_plan(plan) {
    const plan1 = {}

    for (const i in facility_operator_limit) {
      if (i.startsWith('room') && plan[i].name) {
        plan1[i] = remove_empty_agent(plan[i])
      } else {
        let empty = true
        for (const j of plan[i].plans) {
          if (j.agent) {
            empty = false
            break
          }
        }
        if (!empty) {
          plan1[i] = remove_empty_agent(plan[i])
        }
      }
    }

    return plan1
  }

  async function load_plan() {
    const config_store = useConfigStore()
    const { dorm_order } = storeToRefs(config_store)
    // 新排班表重置宿舍优先级
    dorm_order.value = []
    const response = await axios.get(`${import.meta.env.VITE_HTTP_URL}/plan`)
    ling_xi.value = response.data.conf.ling_xi
    exhaust_require.value = str2list(response.data.conf.exhaust_require)
    rest_in_full.value = str2list(response.data.conf.rest_in_full)
    ope_resting_priority.value = str2list(response.data.conf.ope_resting_priority)
    resting_priority.value = str2list(response.data.conf.resting_priority)
    workaholic.value = str2list(response.data.conf.workaholic)
    refresh_trading.value = str2list(response.data.conf.refresh_trading)
    refresh_drained.value = str2list(response.data.conf.refresh_drained)
    const gamings = ['gaming_1', 'gaming_2', 'gaming_3']
    for (const key of gamings) {
      if (!response.data.plan1[key]) {
        response.data.plan1[key] = { plans: [] }
      }
    }
    for (const key of gamings) {
      for (const b of response.data.backup_plans) {
        if (!b.conf[key]) {
          b.conf[key] = { plans: [] }
        }
      }
    }
    plan.value = fill_empty(response.data.plan1)

    backup_plans.value = response.data.backup_plans ?? []
    for (let b of backup_plans.value) {
      for (const i of backup_conf_convert_list) {
        b.conf[i] = str2list(b.conf[i])
      }
      b.plan = fill_empty(b.plan)
    }
  }

  async function load_operators() {
    const response = await axios.get(`${import.meta.env.VITE_HTTP_URL}/operator`)
    const option_list = []
    for (const i of response.data) {
      option_list.push({
        value: i,
        label: i
      })
    }
    operators.value = option_list
  }

  function build_plan() {
    const result = {
      default: 'plan1',
      plan1: strip_plan(plan.value),
      conf: {
        ling_xi: ling_xi.value,
        exhaust_require: list2str(exhaust_require.value),
        rest_in_full: list2str(rest_in_full.value),
        ope_resting_priority: list2str(ope_resting_priority.value),
        resting_priority: list2str(resting_priority.value),
        workaholic: list2str(workaholic.value),
        refresh_trading: list2str(refresh_trading.value),
        refresh_drained: list2str(refresh_drained.value)
      },
      backup_plans: deepcopy(backup_plans.value)
    }
    for (const b of result.backup_plans) {
      for (const i of backup_conf_convert_list) {
        b.conf[i] = list2str(b.conf[i])
      }
      b.plan = strip_plan(b.plan)
    }

    const plan1 = result.plan1

    for (const i in facility_operator_limit) {
      if (i.startsWith('room') && plan.value[i].name) {
        plan1[i] = remove_empty_agent(plan.value[i])
      } else {
        let empty = true
        for (const j of plan.value[i].plans) {
          if (j.agent) {
            empty = false
            break
          }
        }
        if (!empty) {
          plan1[i] = remove_empty_agent(plan.value[i])
        }
      }
    }

    return result
  }

  const loaded = inject('loaded')

  watchEffect(() => {
    if (loaded.value) {
      axios.post(`${import.meta.env.VITE_HTTP_URL}/plan`, build_plan())
    }
  })

  const groups = computed(() => {
    const result = []
    for (const facility in plan.value) {
      for (const p of plan.value[facility].plans) {
        if (p.group) {
          result.push(p.group)
        }
      }
    }
    return [...new Set(result)]
  })

  const sub_plan = ref('main')
  const current_plan = computed(() => {
    if (sub_plan.value == 'main') {
      return plan.value
    } else {
      return backup_plans.value[sub_plan.value].plan
    }
  })

  return {
    load_plan,
    load_operators,
    ling_xi,
    exhaust_require,
    rest_in_full,
    resting_priority,
    ope_resting_priority,
    workaholic,
    refresh_trading,
    refresh_drained,
    plan,
    operators,
    facility_operator_limit,
    left_side_facility,
    build_plan,
    groups,
    backup_plans,
    sub_plan,
    current_plan,
    fill_empty
  }
})

File: arknights_mower/utils/config/plan.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field, model_validator


class MoodLimits(BaseModel):
    lower: float = Field(default=0, ge=0, lt=24, allow_inf_nan=False)
    upper: float = Field(default=24, gt=0, le=24, allow_inf_nan=False)

    @model_validator(mode="after")
    def validate_range(self):
        if self.lower >= self.upper:
            raise ValueError("心情下限必须小于上限")
        return self


class PlanConf(BaseModel):
    ling_xi: int = 1
    "令夕模式，1感知 2烟火 3均衡"
    mood_limits: Optional[MoodLimits] = None
    "全体干员自定义心情上下限；空值沿用自动规则"
    operator_mood_limits: dict[str, MoodLimits] = Field(default_factory=dict)
    "指定干员上下限，优先于全体设置"
    exhaust_require: str = ""
    "耗尽"
    rest_in_full: str = ""
    "回满"
    resting_priority: str = ""
    "低优先级"
    resting_priority_replacement: str = ""
    "测试宿舍逻辑：宿舍高优先级替班，仅提升替班身份"
    resting_standby: str = ""
    "宿舍休息候补干员"
    workaholic: str = ""
    "0心情工作（主力宿舍黑名单）"
    refresh_trading: str = ""
    "跑单时间刷新干员"
    refresh_drained: str = ""
    "用尽时间刷新干员"
    ope_resting_priority: str = ""
    "休息排序优先级"
    dorm_order: str = ""
    "测试宿舍逻辑下当前排班的宿舍房间优先级"


class BackupPlanConf(PlanConf):
    free_blacklist: str = ""
    "（非主力）宿舍黑名单"
    dorm_order_override: Optional[bool] = None
    "是否由该副表显式覆盖此前生效的宿舍房间优先级"


class Plans(BaseModel):
    agent: str
    group: str = ""
    replacement: list[str] = []


class Facility(BaseModel):
    name: str = ""
    plans: list[Plans] = []
    product: Optional[str] = None


class Plan1(BaseModel):
    central: Optional[Facility] = None
    "控制中枢"
    meeting: Optional[Facility] = None
    "会客室"
    factory: Optional[Facility] = None
    "加工站"
    contact: Optional[Facility] = None
    "办公室"
    train: Optional[Facility] = None
    "训练室"
    gaming_1: Optional[Facility] = None
    "活动室1"
    gaming_2: Optional[Facility] = None
    "活动室2"
    gaming_3: Optional[Facility] = None
    "活动室3"
    dormitory_1: Optional[Facility] = None
    dormitory_2: Optional[Facility] = None
    dormitory_3: Optional[Facility] = None
    dormitory_4: Optional[Facility] = None
    room_1_1: Optional[Facility] = None
    room_1_2: Optional[Facility] = None
    room_1_3: Optional[Facility] = None
    room_2_1: Optional[Facility] = None
    room_2_2: Optional[Facility] = None
    room_2_3: Optional[Facility] = None
    room_3_1: Optional[Facility] = None
    room_3_2: Optional[Facility] = None
    room_3_3: Optional[Facility] = None


class Task(BaseModel):
    central: Optional[list[str]] = None
    "控制中枢"
    meeting: Optional[list[str]] = None
    "会客室"
    factory: Optional[list[str]] = None
    "加工站"
    contact: Optional[list[str]] = None
    "办公室"
    train: Optional[list[str]] = None
    "训练室"
    gaming_1: Optional[list[str]] = None
    "活动室1"
    gaming_2: Optional[list[str]] = None
    "活动室2"
    gaming_3: Optional[list[str]] = None
    "训练室3"
    dormitory_1: Optional[list[str]] = None
    dormitory_2: Optional[list[str]] = None
    dormitory_3: Optional[list[str]] = None
    dormitory_4: Optional[list[str]] = None
    room_1_1: Optional[list[str]] = None
    room_1_2: Optional[list[str]] = None
    room_1_3: Optional[list[str]] = None
    room_2_1: Optional[list[str]] = None
    room_2_2: Optional[list[str]] = None
    room_2_3: Optional[list[str]] = None
    room_3_1: Optional[list[str]] = None
    room_3_2: Optional[list[str]] = None
    room_3_3: Optional[list[str]] = None


class Trigger(BaseModel):
    left: str | Trigger = ""
    operator: str = ""
    right: str | Trigger = ""


class BackupPlan(BaseModel):
    conf: BackupPlanConf = {}
    plan: Plan1 = {}
    task: Task = {}
    trigger: Trigger = {}
    trigger_timing: str = "AFTER_PLANNING"
    # 空值表示始终跟随切入时机，兼容旧排班且允许之后修改切入时机。
    exit_trigger_timing: Optional[str] = None
    name: str = "plan"


class PlanModel(BaseModel):
    default: str = "plan1"
    plan1: Plan1 = Plan1()
    conf: PlanConf = PlanConf()
    backup_plans: list[BackupPlan] = []
    # 全局运行设置随排班导出；旧排班没有此字段时保留本机现有设置。
    advanced_settings: Optional[dict[str, Any]] = None


def parse_plan_document(data) -> PlanModel:
    """Reject unrelated JSON instead of silently constructing an empty plan."""
    if not isinstance(data, dict) or not isinstance(data.get("plan1"), dict):
        raise ValueError("排班文件必须包含 plan1 主排班")
    if data.get("default", "plan1") != "plan1":
        raise ValueError("不支持的主排班名称")
    return PlanModel(**data)


def migrate_legacy_dorm_order(
    plan: PlanModel, data: dict, legacy_dorm_order: str
) -> bool:
    """迁移全局旧床位顺序，并折叠为每张排班独立的房间顺序。

    主表缺少独立字段时继承旧全局值；副表只迁移显式的非默认顺序。
    历史版本自动写入副表的 1→2→3→4 视为未覆盖，避免后续副表把
    前一张副表的自定义顺序冲回默认值。
    """
    rooms = [f"dormitory_{index}" for index in range(1, 5)]

    def room_order(value: str) -> str:
        result = []
        for item in (value or "").split(","):
            parts = item.rsplit("_", 1)
            room = (
                parts[0]
                if len(parts) == 2 and parts[0] in rooms and parts[1].isdigit()
                else item
            )
            if room in rooms and room not in result:
                result.append(room)
        result.extend(room for room in rooms if room not in result)
        return ",".join(result)

    changed = False
    main_conf = data.get("conf")
    if not isinstance(main_conf, dict) or "dorm_order" not in main_conf:
        plan.conf.dorm_order = legacy_dorm_order
        changed = True
    normalized = room_order(plan.conf.dorm_order)
    if plan.conf.dorm_order != normalized:
        plan.conf.dorm_order = normalized
        changed = True
    raw_backups = data.get("backup_plans")
    if not isinstance(raw_backups, list):
        raw_backups = []
    for index, backup in enumerate(plan.backup_plans):
        raw_conf = raw_backups[index].get("conf") if index < len(raw_backups) else None
        raw_conf = raw_conf if isinstance(raw_conf, dict) else {}
        raw_order = str(raw_conf.get("dorm_order", "") or "")
        normalized = room_order(raw_order) if raw_order else ""
        explicit = raw_conf.get("dorm_order_override")
        if explicit is None:
            explicit = bool(normalized and normalized != ",".join(rooms))
        explicit = bool(explicit)
        desired_order = normalized if explicit else ""
        if backup.conf.dorm_order != desired_order:
            backup.conf.dorm_order = desired_order
            changed = True
        if backup.conf.dorm_order_override != explicit:
            backup.conf.dorm_order_override = explicit
            changed = True
    return changed
